[Client] Support passing existing executor providers to the client

lhotari opened the pull request:
### Motivation

In load and performance testing, there's a need to simulate production use cases and production workloads.
For this purpose, it would be useful to be able to share the thread pools used by Pulsar client instances in order to be able to run a large amount of Pulsar clients in a single JVM without the overhead of a lot of threads. 
In the current solution, it's already possible to share the EventLoopGroup and HashedWheelTimer instances.
The solution for sharing the thread pools for the external / internal executors was missing. This PR adds support for that.

Example usage:

```java
// shared thread pool related resources
ExecutorProvider internalExecutorProvider = new ExecutorProvider(8, "shared-internal-executor");
ExecutorProvider externalExecutorProvider = new ExecutorProvider(8, "shared-external-executor");
Timer sharedTimer = new HashedWheelTimer(getThreadFactory("shared-pulsar-timer"), 1, TimeUnit.MILLISECONDS);
EventLoopGroup sharedEventLoopGroup = new EpollEventLoopGroup();

// example of creating a client which uses the shared thread pools
PulsarClientImpl client = PulsarClientImpl.builder().conf(conf)
                .internalExecutorProvider(internalExecutorProvider)
                .externalExecutorProvider(externalExecutorProvider)
                .timer(sharedTimer)
                .eventLoopGroup(sharedEventLoopGroup)
                .build();
```

It seems that this would also improve the performance of the Pulsar Proxy since new thread pools for every client connection.

That happens in the Pulsar Proxy currently: 
https://github.com/apache/pulsar/blob/af63e96d4aaa0ae4c4086583aa4f9b1edd72279b/pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java#L445-L451

An optimization was added in #9802 for sharing the timer, but it would be useful to also share the internal / external executors.


### Modifications

Refactor the code to add an internal builder class so that more constructors don't have to be added for advanced initialization of the PulsarClientImpl class.

eolivelli commented on pulsar-client/src/test/java/org/apache/pulsar/client/impl/PulsarClientImplTest.java:
```diff
@@ -217,4 +219,32 @@ public void testResourceCleanup() throws PulsarClientException {
+                .build();
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
```
we should add a more narrow verification of the assertion, IllegalStateException may be thrown for other reasons

lhotari replied:
Fixed

eolivelli reviewed (changes requested):
While I appreciate the feature and I see the benefits,
it looks like we are adding a new "official" way to build a Pulsar Client, that is not via the public API.

Do you want to use this only inside internal Pulsar tools ? the Proxy, Pulsar perf, Pulsar CLI tools ?

Why can't we add this support using the regular PulsarClient Builder API ?

lhotari commented:
> Do you want to use this only inside internal Pulsar tools ? the Proxy, Pulsar perf, Pulsar CLI tools ?

yes, mainly an internal, unsupported API. Pulsar Proxy and Pulsar perf are the main internal use cases.

> Why can't we add this support using the regular PulsarClient Builder API ?

It seems that this would just add unnecessary complexity and extra baggage in the public API. 
I'd assume that there are very few use cases where there would be a need to specify the thread pools to use.

My use case is in improving [nosqlbench's Pulsar driver](https://github.com/nosqlbench/nosqlbench/tree/main/driver-pulsar). This would allow to emulate thousands of isolated Pulsar clients within a single JVM. When a Pulsar client is shared, it doesn't properly emulate real production work loads.

Anonymitaet commented:
@lhotari Thanks for your contribution. For this PR, do we need to update docs?

(The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc-related info in this and future PR descriptions? Thanks)

315157973 commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java:
```diff
@@ -133,48 +131,53 @@ public SchemaInfoProvider load(String topicName) {
+            connectionPoolReference = connectionPool != null ? connectionPool : new ConnectionPool(conf, this.eventLoopGroup);
+            this.cnxPool = connectionPoolReference;
+            this.externalExecutorProvider = externalExecutorProvider != null ? externalExecutorProvider : new ExecutorProvider(conf.getNumListenerThreads(), "pulsar-external-listener");
+            this.internalExecutorProvider = internalExecutorProvider != null ? internalExecutorProvider : new ExecutorProvider(conf.getNumIoThreads(), "pulsar-client-internal");
```
In my opinion, I don't think that sharing the internalExecutorProvider is a good way. This thread pool is very critical. It is best to isolate the thread pool. If any tasks that take a long time are executed in it, it will affect the consumption performance of the client.
This thread pool is now shared by all Producers and Consumers. Usually when we use the SDK, one process will only create one PulsarClient instance.There will not be too many threads to create.

lhotari replied:
@315157973  I agree with you. Please check the details of my use case. That's what makes the difference.

lhotari replied:
My use case is in improving [nosqlbench's Pulsar driver](https://github.com/nosqlbench/nosqlbench/tree/main/driver-pulsar). This would allow to emulate thousands of isolated Pulsar clients within a single JVM. When a Pulsar client is shared, it doesn't properly emulate real production work loads.

lhotari replied:
The main reason why it doesn't emulate production work loads is the connection pool and multiplexing over TCP/IP connections that the Pulsar client does for producers and consumers. When emulating production workloads in load testing, that's something that makes a big difference.

lhotari replied:
@315157973 The other important use case is the Pulsar Proxy. Please check the PR description for details.

315157973 replied:
OK, I get your point

codelipenghui commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java:
```diff
@@ -133,48 +131,53 @@ public SchemaInfoProvider load(String topicName) {
+            this.createdEventLoopGroup = eventLoopGroup == null;
+            this.createdCnxPool = connectionPool == null;
+            if ((externalExecutorProvider == null) != (internalExecutorProvider == null)) {
+                throw new IllegalStateException("Both externalExecutorProvider and internalExecutorProvider must be specified or unspecified.");
```
It should be IllegalArgumentException

lhotari replied:
Fixed

codelipenghui commented:
@lhotari Could you please help resolve the conflicts? It's a nice have in 2.10.

lhotari commented:
> @lhotari Could you please help resolve the conflicts? It's a nice have in 2.10.

@codelipenghui I have rebased the changes and addressed the review comments. PTAL

lhotari commented:
@codelipenghui This PR is now ready for final review
